### PR TITLE
Repackage poly into subpackages.

### PR DESCRIPTION
**Pull request by TimothyStiles**
Poly has gotten really big and it makes sense to repackage a lot of it into subpackages.

This PR does exactly that. It runs and all tests pass. If anyone could review it for syntax, clarity, and overall project structure that'd be great.

Before merging I'm going to work on adding go docs for each subpackage with runable examples that will sort of function like integration tests where appropriate and give end users an idea of how all this stuff ties together.

-Tim

EDIT: Should also find a way to make test data more manageable without duplication.

### CODE REVIEW

**Review by isaacguerreir — APPROVED** [review]
Didn't found any problem.

The code is more readable, definitly. Thanks, Tim!

File: io/genbank/genbank_test.go
Comment (TimothyStiles) [inline io/genbank/genbank_test.go:36]: Should just be ExampleBuild()

File: io/pson/pson.go
Comment (TimothyStiles) [inline io/pson/pson.go:36]: *writes a*

File: io/genbank/genbank.go
Comment (TimothyStiles) [inline io/genbank/genbank.go:96]: should be renamed to `Build`.

File: clone/clone.go
Comment (jkhales) [inline clone/clone.go:61]:
```suggestion
type DNASequence struct {
```
Part is maybe too generic

File: docs/library-io.md
Comment (jkhales) [inline docs/library-io.md:176]: There is a `BuildGff` right above this. Naming is inconsistent/confusing.

Maybe make a universal `Build(..., 'GBK')` interface?

File: transform/reverse/reverse_test.go
Comment (jkhales) [inline transform/reverse/reverse_test.go:1]: Either add tests or delete this?

File: transform/codon/codon.go
Comment (jkhales) [inline transform/codon/codon.go:75]: `Table` seems less descriptive and grep'able than `CodonTable`.

**Review by jkhales — APPROVED** [review]
✨

File: docs/library-io.md
Comment (TimothyStiles) [reply docs/library-io.md:176]: I was actually going to remove these docs for good with this PR. Too likely to break and not helpful. Should be replaced with new tutorials that are coming in new PR.

> Maybe make a universal Build(..., 'GBK') interface?

Possibly but I really like the current interface which is `genbank.Build(Sequence)`

File: transform/codon/codon.go
Comment (TimothyStiles) [reply transform/codon/codon.go:75]: That's true but when used in this new subpackage configuration it's written as `codon.Table` in packages that depend on it so it'd likely be just as grepable. Does feel super weird in package though. I've got similar feelings about `random.ProteinSequence`.

File: clone/clone.go
Comment (TimothyStiles) [reply clone/clone.go:61]: This was on @Koeng101's suggestion. It was originally called `CloneSequence` but would end up being written as `clone.CloneSequence` when used from another package. I couldn't have it be named `clone.Sequence` since it'd get confusing because the main poly package contains the `Sequence` struct for annotated sequences. I suggested `clone.Vector` but semantically `clone.Part` is most correct. I'll add some DNA part terminology to the comments to clarify this for future devs.

File: transform/reverse/reverse_test.go
Comment (TimothyStiles) [reply transform/reverse/reverse_test.go:1]: I'll add tests. That was on my todo list thanks for catching this and reminding me.

Comment (TimothyStiles) [reply transform/reverse/reverse_test.go:1]: Just added a test. Done!

File: clone/clone.go
Comment (Koeng101) [reply clone/clone.go:61]: I think Part is the most correct domain-wise, but I think jkhales is right that it is too generic for folks not building DNA. I would say keep Part but add a much larger comment.

File: transform/reverse/reverse_test.go
Comment (TimothyStiles) [reply transform/reverse/reverse_test.go:1]: Wait no. Test failed. That's what I get for using GitHub copilot. Was just excited that it wrote the test for me 😅

Comment (Koeng101) [reply transform/reverse/reverse_test.go:1]: Should this just be in a general utils package?

File: transform/codon/codon.go
Comment (Koeng101) [reply transform/codon/codon.go:75]: Yea probably fine now that we have subpackages.

Comment (jkhales) [reply transform/codon/codon.go:75]: ah, if it's always written out as `codon.Table` that is pretty nice.

File: checks/checks.go
Comment (Koeng101) [inline checks/checks.go:9]: I think we might want to merge this into a utils package with basic complementation and reverse stuff. Thoughts?

Comment (Koeng101) [reply checks/checks.go:9]: Also I don't know how to fix this, but if I remember correctly you code copy paste rather than the copy-paste-file then remove-content, you remove the original contributors "ownership" or whatever git tracks. I don't care, but other folks might.

File: io/example_test.go
Comment (Koeng101) [inline io/example_test.go:7]: What is pson?

File: transform/reverse/reverse_test.go
Comment (jkhales) [reply transform/reverse/reverse_test.go:1]: I'm jealous, extremely impatiently waiting on copilot access.

image

File: io/pson/pson.go
Comment (Koeng101) [inline io/pson/pson.go:1]: Is PSON an alternative to JSON or something? P isn't really around J so I'm wondering if this is not a typo.

File: transform/variants/variants.go
Comment (Koeng101) [inline transform/variants/variants.go:12]: Did this remove contribution?

File: transform/reverse/reverse_test.go
Comment (TimothyStiles) [reply transform/reverse/reverse_test.go:1]: > Should this just be in a general utils package?

Go frowns upon this practice a lot and I tend to agree. Easy to dump a bunch of unrelated stuff in there. I do think the naming for this should be a lot smarter cause it turns out that co-pilot actually wrote a test for a function that returns a complement which without the package name is pretty much what `reverse.Complement` looks like.

File: io/example_test.go
Comment (TimothyStiles) [reply io/example_test.go:7]: `poly json` thought I wrote a comment about it but essentially the namespace collisions with Go's `json` package was too much of a pain. Should make a comment if there isn't already

File: transform/reverse/reverse_test.go
Comment (Koeng101) [reply transform/reverse/reverse_test.go:1]: ok, that's fair! If it's the go way, might as well

File: io/pson/pson.go
Comment (TimothyStiles) [reply io/pson/pson.go:1]: > poly json thought I wrote a comment about it but essentially the namespace collisions with Go's json package was too much of a pain. Should make a comment if there isn't already

File: transform/variants/variants.go
Comment (TimothyStiles) [reply transform/variants/variants.go:12]: It shouldn't have? I think historically it has done that but git seemed to be tracking where code was going. I'll look into it.

File: checks/checks.go
Comment (TimothyStiles) [reply checks/checks.go:9]: > I think we might want to merge this into a utils package with basic complementation and reverse stuff. Thoughts?

See -> https://github.com/TimothyStiles/poly/pull/174#discussion_r661935919

Comment (TimothyStiles) [reply checks/checks.go:9]: > Also I don't know how to fix this, but if I remember correctly you code copy paste rather than the copy-paste-file then remove-content, you remove the original contributors "ownership" or whatever git tracks. I don't care, but other folks might.

I worried about this but GitHub seemed to be tracking this pretty well and it may be a new feature of git. I'll look into this before merge.

File: transform/reverse/reverse_test.go
Comment (TimothyStiles) [reply transform/reverse/reverse_test.go:1]: Also tricky thing about `reverse.complement` is that it's used almost everywhere and was causing serious cyclic dependency issues to the point where it kind of needed its own package that didn't rely on anything else. This is probably the case with a lot of util like functions.

File: io/pson/pson.go
Comment (TimothyStiles) [reply io/pson/pson.go:1]: See -> https://github.com/TimothyStiles/poly/pull/174#discussion_r661936455

File: io/example_test.go
Comment (jkhales) [reply io/example_test.go:7]: FWIW, `pson` is definitely a clever name, but `polyJson` maybe a clearer name and avoid confusion such as experienced by @Koeng101 I also didn't realize it was a json wrapper.

Comment (Koeng101) [reply io/example_test.go:7]: I think polyJson would be better

File: transform/reverse/reverse_test.go
Comment (TimothyStiles) [reply transform/reverse/reverse_test.go]: Just fixed this in latest push. I moved `ReverseComplement` to the `transform` package so now you call it with `transform.ReverseComplement`. I also pushed `Reverse` and `Complement` functions and Example tests. @Koeng101 @jkhales

File: io/example_test.go
Comment (TimothyStiles) [reply io/example_test.go]: > I think polyJson would be better

@Koeng101 @jkhales Just changed to polyjson in latest push.

**Comment by TimothyStiles** [comment]
@Koeng101 @jkhales @isaacguerreir I'm planning to merge this by the end of today if there's no major objections. I looked into and it seems that I'll have to make a new PR to make sure git blame doesn't get borked and all previous contributors work will be credited but otherwise I think this is ready for a final review.

Also, I've asked @Koeng101 to write an explanation of DNA parts for `clone.Part` as requested by @jkhales that will be included in a new pull request.

**Review by Koeng101 — APPROVED** [review]
I approve of this repackage. From my look at the files, I don't see anything that is off.

**Pull request merged**
